File: backend/core/service/game_service.py
import logging
from math import sqrt
from random import sample
from typing import List

from django_eventstream import send_event

from core.messages import (
    DebriefStartsMessage,
    RoundStartsMessage,
    VoteResultsStartsMessage,
)
from core.models import (
    Game,
    GamePhase,
    Pad,
    PadStep,
    Player,
    PlayerGameParticipation,
    Room,
    StepType,
)

logger = logging.getLogger(__name__)


def get_round_count(game: Game):
    player_count = game.participants.count()
    # There should always be an even number of steps - the first player will play a step on their pad
    # or won't depending on that
    return 2 * ((player_count + int(game.draw_own_word) - 1) // 2)


def order_players(players: List[Player], requested_players_order: List[str]):
    if requested_players_order is None:
        return sample(players, len(players))

    def sort_fn(player: Player):
        try:
            return requested_players_order.index(player.uuid.hex)
        except ValueError:
            return -1

    return sorted(players, key=sort_fn)


def get_available_vote_count(game: Game):
    round_count = get_round_count(game)
    choice_count = round_count * (game.participants.count() - 1)
    return max(1, round(sqrt(0.6 * choice_count - 1)))


def initialize_game(
    room: Room,
    requested_players_order: List[str],
    round_duration: int,
    draw_own_word: bool,
):
    logger.debug("Initializing game for room %s" % room.uuid)

    players = list(room.players.all())
    player_count = len(players)
    even_players = player_count % 2 == 0

    game = Game.objects.create(
        room=room,
        round_duration=round_duration,
        draw_own_word=(player_count == 2 or (even_players and draw_own_word)),
    )

    ordered_players = order_players(players, requested_players_order)

    for index, player in enumerate(ordered_players):
        PlayerGameParticipation.objects.create(game=game, player=player, order=index)

    for index in range(len(ordered_players)):
        initialize_pad(game, index, ordered_players)

    return game


def initialize_pad(game: Game, index: int, players: List[Player]):
    logger.debug(
        "Initializing pad for player %s (index %d)" % (players[index].name, index)
    )
    pad = Pad.objects.create(game=game, initial_player=players[index], order=index)
    player_count = len(players)

    step_count = get_round_count(game)

    for round_number in range(step_count):
        step_player = (
            players[(index + round_number) % player_count]
            if game.draw_own_word
            else players[(index + round_number + 1) % player_count]
        )
        step_type = (
            StepType.WORD_TO_DRAWING
            if round_number % 2 == 0
            else StepType.DRAWING_TO_WORD
        )
        logger.debug(
            "Initializing pad step for pad %s, round %d, with player %s and mode %s"
            % (pad.uuid.hex[:8], round_number, step_player.name, step_type.value)
        )

        PadStep.objects.create(
            pad=pad,
            player=step_player,
            round_number=round_number,
            step_type=step_type.value,
        )


def switch_to_rounds(game: Game):
    game.phase = GamePhase.ROUNDS.value
    game.current_round = 0
    game.pads_done = 0
    game.save()
    for pad in game.pads.all():
        step = pad.steps.get(round_number=game.current_round)
        step.sentence = pad.sentence
        step.save()

    send_event(
        "game-%s" % game.uuid.hex, "message", RoundStartsMessage(game, 0).serialize(),
    )


def start_next_round(game: Game, new_round: int):
    round_count = get_round_count(game)
    if new_round >= round_count:
        return start_debrief(game)
    game.current_round = new_round
    game.pads_done = 0
    game.save()
    for pad in game.pads.all():
        previous_step = pad.steps.get(round_number=new_round - 1)
        step = pad.steps.get(round_number=new_round)
        if new_round % 2 == 0:
            step.sentence = previous_step.sentence
        else:
            step.drawing = previous_step.drawing
        step.save()

    send_event(
        "game-%s" % game.uuid.hex,
        "message",
        RoundStartsMessage(game, round_number=new_round).serialize(),
    )


def switch_to_vote_results(game: Game):
    game.phase = GamePhase.VOTE_RESULTS.value
    game.save()
    send_event(
        "game-%s" % game.uuid.hex,
        "message",
        VoteResultsStartsMessage(game).serialize(),
    )


def start_debrief(game: Game):
    game.phase = GamePhase.DEBRIEF.value
    game.current_round = None
    game.pads_done = 0
    game.save()
    send_event(
        "game-%s" % game.uuid.hex, "message", DebriefStartsMessage(game).serialize(),
    )


class GamePhaseAssertionException(Exception):
    def __init__(self, expected, actual):
        message = "Expected game to be at phase %s, actual phase %s" % (
            expected,
            actual,
        )
        super(Exception, self).__init__(message)
        self.expected = expected
        self.actual = actual


class GameRoundAssertionException(Exception):
    def __init__(self, expected, actual):
        message = "Expected game to be at round %d, actual round %d" % (
            expected,
            actual,
        )
        super(Exception, self).__init__(message)
        self.expected = expected
        self.actual = actual


def assert_phase(game: Game, expected_phase: GamePhase):
    if game.phase != expected_phase.value:
        raise GamePhaseAssertionException(expected_phase.value, game.phase)


def assert_round(game: Game, expected_round: int):
    if game.current_round != expected_round:
        raise GameRoundAssertionException(expected_round, game.current_round)


def is_player_in_game(game: Game, player: Player):
    return player in [participant.player for participant in game.participants.all()]

File: frontend/src/services/game.service.ts
import { Game, GamePhase } from 'redux/Game/types';
import { Player } from 'redux/Player/types';
import { Room } from 'redux/Room/types';

export const getRedirectPath = (room: Room, game: Game, player: Player) => {
  switch (game.phase) {
    case GamePhase.INIT:
      const playerPad = game.pads.find((pad) => pad.initial_player.uuid === player.uuid);
      if (!playerPad) {
        throw new Error(`Pad for player ${player.uuid} not found in game ${game.uuid}`);
      }
      return `/room/${room.uuid}/game/${game.uuid}/pad/${playerPad.uuid}/init`;
    case GamePhase.ROUNDS:
      const playerStep = game.rounds.find(
        (step) => step.player.uuid === player.uuid && step.round_number === game.current_round,
      );
      if (!playerStep) {
        throw new Error(
          `Step for player ${player.uuid} and round ${game.current_round} not found in game ${game.uuid}`,
        );
      }
      return `/room/${room.uuid}/game/${game.uuid}/step/${playerStep.uuid}`;
    case GamePhase.DEBRIEF:
      return `/room/${room.uuid}/game/${game.uuid}/recap`;
    case GamePhase.VOTE_RESULTS:
      return `/room/${room.uuid}/game/${game.uuid}/vote-results`;
  }
};

export const getAvailableVoteCount = (game: Game): number => {
  const stepCount = game.pads[0].steps.length;
  const playerCount = game.players.length;
  const choiceCount = stepCount * (playerCount - 1);
  return Math.max(1, Math.round(Math.sqrt(0.6 * choiceCount - 1)));
};

export const getReorderedPlayers = (game: Game, player: Player): Player[] => {
  const currentPlayerPad = game.pads.find((pad) =>
    game.phase === GamePhase.INIT
      ? pad.initial_player.uuid === player.uuid
      : pad.steps.some(
          (step) => step.round_number === game.current_round && step.player.uuid === player.uuid,
        ),
  );
  if (!currentPlayerPad) {
    throw new Error(
      `Step for player ${player.uuid} and round ${game.current_round} not found in game ${game.uuid}`,
    );
  }

  return [currentPlayerPad.initial_player, ...currentPlayerPad.steps.map((step) => step.player)];
};

export const getNextPhaseAndRound = (game: Game) => {
  switch (game.phase) {
    case GamePhase.INIT:
      return [GamePhase.ROUNDS, 0];
    case GamePhase.ROUNDS: {
      const nextRoundNumber = (game.current_round || 0) + 1;
      const roundCount = game.pads[0].steps.length;

      if (nextRoundNumber >= roundCount) {
        return [GamePhase.DEBRIEF, 0];
      }
      return [GamePhase.ROUNDS, nextRoundNumber];
    }
    case GamePhase.DEBRIEF:
    case GamePhase.VOTE_RESULTS:
      return [GamePhase.VOTE_RESULTS, 0];
  }
};

export const shouldDisplayDrawOwnWordSwitch = (playerCount: number) => {
  if (playerCount <= 3) {
    return false;
  }
  return playerCount % 2 === 0;
};
